### DISCUSSION TITLE
Database+Bearer authentication continuously unauthorized

### DISCUSSION THREAD

**Discussion by mayukhnair** [fastapi-users/fastapi-users#987] [Q&A] [answered]
My authentication backend is database + bearer. Requests being sent with a Bearer authorization in the header randomly seems to be rejected with a 401 Unauthorized or passed, the former much more frequently than the latter.

This is happening even after I set `lifetime_seconds` to `None` or 3.12e7 seconds (equivalent to a year).

```
class SimpleBearerTransport(BearerTransport):
    def __init__(self):
        self.scheme = HTTPBearer()  # type: ignore

class UserManager(BaseUserManager[UserCreate, UserDB]):
    user_db_model = UserDB
    reset_password_token_secret = SECRET
    verification_token_secret = SECRET

    async def on_after_register(self, user: UserDB, request: Optional[Request] = None):
        print(f"User {user.id} has registered.")

    async def on_after_forgot_password(
        self, user: UserDB, token: str, request: Optional[Request] = None
    ):
        # TODO: Email + web flow for forgotten passwords
        print(f"User {user.id} has forgot their password. Reset token: {token}")

    async def on_after_request_verify(
        self, user: UserDB, token: str, request: Optional[Request] = None
    ):
        # TODO: Email flow for user verification
        print(f"Verification requested for user {user.id}. Verification token: {token}")


async def get_user_manager(user_db: SQLAlchemyUserDatabase = Depends(get_user_db)):
    yield UserManager(user_db)


bearer_transport = SimpleBearerTransport()

def get_database_strategy(
    access_token_db: AccessTokenDatabase[AccessToken] = Depends(get_access_token_db),
) -> DatabaseStrategy[UserCreate, UserDB, AccessToken]:
    return DatabaseStrategy(access_token_db, lifetime_seconds=None)

auth_backend = AuthenticationBackend(
    name="database",
    transport=bearer_transport,
    get_strategy=get_database_strategy,
)
fastapi_users = FastAPIUsers(
    get_user_manager,
    [auth_backend],
    User,
    UserCreate,
    UserUpdate,
    UserDB,
)

current_active_user = fastapi_users.current_user(active=True)
```

The idea for my flow is that once the user logs in or signs up with a username and password, they immediately get a bearer token (and the verification token as well for the latter). Verified & active users can use the token in any call to request the data. The token should be valid for all calls up to a year without requiring to log in again.

This issue is even affecting newly registered and verified users with no fix in sight.

```
# Bearer token belongs to a user registered and verified 5 minutes ago, should suffice under the default 3600 second lifetime_seconds
(env) mayukh@DESKTOP:~/api$ curl --location --request GET 'http://localhost:8000/agmarknet/get-commodities?type=1' --header 'Authorization: Bearer <bearer token>'
{"detail":"Unauthorized"}
```

Additional info:

```
fastapi==0.75.2
fastapi-users==9.3.1
fastapi-users-db-sqlalchemy==3.0.1

```

**Comment by frankie567** [comment]
Could you try to look directly into your database and check that the access token is properly created with a valid `created_at` value?

**Reply by mayukhnair** [reply]
Can confirm all access tokens are being correctly formed with a valid `created_at` value (YYYY-MM-DD HH:MM:SS) and assigned to an appropriate user ID.

**Reply by frankie567** [reply]
Ok, I suspect a timezone issue that I fixed in `fastapi-users-db-sqlalchemy>=4.0.3`. I'll backport it for `v3` branch.

**Reply by mayukhnair** [reply]
Great, thanks. Or should I port everything to the latest `fastapi` & `fastapi-users`? Would that be the better approach?

**Reply by frankie567** [reply]
I would indeed recommend to migrate to the latest FastAPI Users version. However, I understand that it takes time, so I won't that it's *absolutely* necessary to do it *right now* 😉

Anyway, I released [fastapi-users-db-sqalchemy==3.0.2](https://pypi.org/project/fastapi-users-db-sqlalchemy/3.0.2/) with the timezone fix. Try to install it and tell me if it solves the issue :)

**Reply by mayukhnair** [reply]
Hi, I just tried the updated library and it still has the same error. I will now try porting the code to work with both libraries and see how things work out.

**Reply by mayukhnair** [reply]
You might want to try out the database+bearer auth backend for the previous version though? As you said, this will actually take some time :)

**Reply by frankie567** [reply]
Ok so after trying, I understood what the issue is. Actually, it comes from the `SimpleBearerTransport` trick I suggested. The problem when using the [HTTPBearer scheme is that it returns an `HTTPAuthorizationCredentials` object](https://github.com/tiangolo/fastapi/blob/2aaac141ddb79d1f9071f8e02d2b2097f3d4589f/fastapi/security/http.py#L113-L133), which contains both the scheme and the token.

On the other hand, [FastAPI Users Authenticator expects a plain string](https://github.com/fastapi-users/fastapi-users/blob/b7734fc8b05b10507bc7f29afde1e15956fefe9d/fastapi_users/authentication/authenticator.py#L165-L172), not this kind of object; which causes the error (contrary to you, I had a proper Internal Server Error with the full stack-trace which helped me understand — you maybe have a middleware or dependency swallowing the errors —).

So, the solution could be to overload `HTTPBearer` to manually explode the `HTTPAuthorizationCredentials` object:

```py
class HTTPBearerTokenOnly(HTTPBearer):
    async def __call__(self, request: Request) -> Optional[str]:  # type: ignore
        credentials = await super().__call__(request)
        if credentials:
            return credentials.credentials
        return None


class SimpleBearerTransport(BearerTransport):
    def __init__(self):
        self.scheme = HTTPBearerTokenOnly()  # type: ignore
```

A bit complicated, but working.

**Reply by mayukhnair** [reply]
Okay so this indeed did manage to work! Thank you for the help. 

I will be linking this in the original question about `SimpleBearerTransport` as well to warn future viewers. Also, it would be great if this becomes an included feature in `fastapi-users`. Again, thanks for all the kind, patient help!

**Reply by mayukhnair** [reply]
Also, can you edit your answer with this code block so I can mark it as answered?

**Reply by frankie567** [reply]
Great! Sure, I'll do this!

**Comment by mayukhnair** [comment, marked as answer]
Solution courtesy @frankie567 in the comments. I can verify this works.

> Ok so after trying, I understood what the issue is. [Actually, it comes from the `SimpleBearerTransport` trick I suggested](https://github.com/fastapi-users/fastapi-users/discussions/973). The problem when using the [HTTPBearer scheme is that it returns an `HTTPAuthorizationCredentials` object](https://github.com/tiangolo/fastapi/blob/2aaac141ddb79d1f9071f8e02d2b2097f3d4589f/fastapi/security/http.py#L113-L133), which contains both the scheme and the token.
> 
> On the other hand, [FastAPI Users Authenticator expects a plain string](https://github.com/fastapi-users/fastapi-users/blob/b7734fc8b05b10507bc7f29afde1e15956fefe9d/fastapi_users/authentication/authenticator.py#L165-L172), not this kind of object; which causes the error (contrary to you, I had a proper Internal Server Error with the full stack-trace which helped me understand — you maybe have a middleware or dependency swallowing the errors —).
> 
> So, the solution could be to overload `HTTPBearer` to manually explode the `HTTPAuthorizationCredentials` object:
> 
> ```python
> class HTTPBearerTokenOnly(HTTPBearer):
>     async def __call__(self, request: Request) -> Optional[str]:  # type: ignore
>         credentials = await super().__call__(request)
>         if credentials:
>             return credentials.credentials
>         return None
> 
> 
> class SimpleBearerTransport(BearerTransport):
>     def __init__(self):
>         self.scheme = HTTPBearerTokenOnly()  # type: ignore
> ```
> 
> A bit complicated, but working.
